echopop/computation/biology.py

```diff
@@ -104,160 +128,751 @@ def sum_strata_weight( haul_data: pd.DataFrame ,
-        .dropna( how = 'any' )
-        .loc[ lambda x: x.sex != 'unsexed' ]
-        .pipe( lambda df: pd.concat( [ df , 
-                                       df.copy( ).assign( sex = 'all' ) ] ) )
+    # ---- Bin length and age measurements
+    specimen_data_filtered = (
+        specimen_data_filtered
+        # ---- Bin length
+        .bin_variable( length_intervals , 'length' )
+        # ---- Age bin
         .bin_variable( age_intervals , 'age' )
-        .bin_variable( length_intervals , 'length' )      
-        .groupby( [ 'stratum_num' , 'age_bin' , 'length_bin' , 'sex' ] )
-        .apply( lambda df: pd.Series( { 'summed_weight_all': df.weight.sum( ) ,
-                                        'summed_weight_adult': df.loc[ df.age > 1 ].weight.sum( ) } ) ) 
+    )
+
+    ### Sum weights within each length and age bin for each sex within each stratum
+    # ---- Create separate 'weight_adult' column
+    specimen_data_filtered[ 'weight_adult' ] = (
+        np.where( specimen_data_filtered.age > 1 , specimen_data_filtered.weight , 0.0 )
+    )
+    
+    # ---- Calculate aggregate sums of fish weights
+    specimen_binned_weight = (
+        specimen_data_filtered
+        # ---- Group weight summations across stratum/species/sex/length/age
+        .groupby( [ 'stratum_num' , 'species_id' , 'sex' , 'length_bin' , 'age_bin' ] )
+        # ---- Sum the weights 
+        .agg( weight_all =( 'weight' , 'sum' )  , weight_adult =( 'weight_adult' , 'sum' ) ) 
+        # ---- Fill empty/non-existent values with 0's
+        .fillna( 0 )
         .reset_index( )
-        .replace( np.nan , 0 )
-        .assign( total_weight_sex_all = lambda df: df.groupby( [ 'stratum_num' , 'sex' ] )[ 'summed_weight_all' ].transform( sum ) ,
-                 total_weight_sex_adult = lambda df: df.groupby( [ 'stratum_num' , 'sex' ] )[ 'summed_weight_adult' ].transform( sum ) ,
-                 proportion_weight_all = lambda df: df.summed_weight_all / df.total_weight_sex_all ,
-                 proportion_weight_adult = lambda df: df.summed_weight_adult / df.total_weight_sex_adult )            
+    )
+    
+    ### Calculate the relative weight proportions of each length-age bin for each sex within each stratum
+    proportions_weight_length_age_sex = (
+        specimen_binned_weight
+        # ---- Calculate total sex-specific weights for each stratum
+        .assign( total_weight_sex_all = lambda df: df.groupby( [ 'stratum_num' , 'species_id' , 'sex'  ] )[ 'weight_all' ].transform( sum ) ,
+                 total_weight_sex_adult = lambda df: df.groupby( [ 'stratum_num' , 'species_id' , 'sex' ] )[ 'weight_adult' ].transform( sum ) )
+        # ---- Calculate the weight proportions within each sex: from Matlab --> Len_age_key_wgt_*n
+        .assign( proportion_weight_sex_all = lambda df: df.weight_all / df.total_weight_sex_all ,
+                 proportion_weight_sex_adult = lambda df: df.weight_adult / df.total_weight_sex_adult )
```

Comment:
It'll probably be a good idea to change `weight_all` to something like `weight_sex_length_age_all` so that it is clear that these are grouped separately, for readability.